# 01 - Exploratory Data Analysis

**Dataset:** Amazon Reviews 2023 — Video_Games category

**Files:** data/raw/video_games_reviews_100000.parquet, data/raw/video_games_metadata_100000.parquet

**Goal:** Understand data quality and distribution to inform cleaning/filtering decisions in src/data/clean.py and src/data/split.py

## 1. Load and Inspect Raw Data

In [4]:
import pandas as pd

reviews_df = pd.read_parquet("../data/raw/Video_Games_reviews_100000.parquet")
metadata_df = pd.read_parquet("../data/raw/Video_Games_metadata_100000.parquet")

print(reviews_df.shape, metadata_df.shape)
print(reviews_df.columns.tolist())
print(metadata_df.columns.tolist())

(3394737, 10) (100000, 16)
['rating', 'title', 'text', 'images', 'asin', 'parent_asin', 'user_id', 'timestamp', 'helpful_vote', 'verified_purchase']
['main_category', 'title', 'average_rating', 'rating_number', 'features', 'description', 'price', 'images', 'videos', 'store', 'categories', 'details', 'parent_asin', 'bought_together', 'subtitle', 'author']


In [5]:
reviews_df.head()

,rating,title,text,images,asin,parent_asin,user_id,timestamp,helpful_vote,verified_purchase
0,4.0,It’s pretty sexual. Not my fav,I’m playing on ps5 and it’s interesting. It’s...,[],B07DJWBYKP,B07DK1H3H5,AGCI7FAH4GL5FI65HYLKWTMFZ2CQ,1608186804795,0,True
3,5.0,Great alt to pro controller,"These work great, They use batteries which is ...",[],B07GXJHRVK,B0BCHWZX95,AFTC6ZR5IKNRDG5JCPVNVMU3XV2Q,1577637634017,0,True
4,5.0,solid product,I would recommend to anyone looking to add jus...,[],B00HUWA45W,B00HUWA45W,AFTC6ZR5IKNRDG5JCPVNVMU3XV2Q,1427591932000,0,True
5,3.0,love all the amazing colors but the black is r...,love all the amazing colors but the black is r...,[],B016Y2BVKA,B073SC6V1D,AHXSBZT52TCPZUBVCBRICTHWUCBA,1518124539574,0,True
6,5.0,Will use again,Instant delivery!,[],B004RMK57U,B004RMK57U,AHZIJGKEWRTAEOZ673G5B3SNXEGQ,1602937709361,0,True


In [6]:
metadata_df.head()

,main_category,title,average_rating,rating_number,features,description,price,images,videos,store,categories,details,parent_asin,bought_together,subtitle,author
0,Video Games,WRC 7 - PlayStation 4,4.2,161,[New official cars! Powerful rally behemoths f...,"[The racing game dominating the genre is back,...",None,"{'hi_res': [None, 'https://m.media-amazon.com/...","{'title': ['WRC 7 - PlayStation 4'], 'url': ['...",Maximum Games,"[Video Games, PlayStation 4, Games]","{""Release date"": ""October 3, 2017"", ""Best Sell...",B0754RPTB7,NaN,NaN,NaN
1,Video Games,"TOPAD Remote for PS4 Controller,Gamepad Joysti...",3.8,41,[🎮【Ergonomic Design & Comfortable】TOPAD ymir b...,[],24.99,{'hi_res': ['https://m.media-amazon.com/images...,"{'title': ['Bang for Buck', 'Awesome', 'Great ...",TOPAD,"[Video Games, PlayStation 4, Accessories, Remo...","{""Best Sellers Rank"": {""Video Games"": 27896, ""...",B0BW6QSLXM,NaN,NaN,NaN
2,Video Games,"GEEKERA Charging Dock for Oculus Quest 2, Magn...",4.5,73,[Effortless Magnetic Charging: Experience hass...,[],48.99,{'hi_res': ['https://m.media-amazon.com/images...,"{'title': [], 'url': [], 'user_id': []}",GEEKERA,"[Video Games, PC, Virtual Reality, Headsets]","{""Best Sellers Rank"": {""Video Games"": 2202, ""P...",B0BZD7TTVZ,NaN,NaN,NaN
3,All Electronics,Switch Lite Case - innoAura 15 in 1 Switch Lit...,4.8,848,[【15 in 1 NS Switch Lite Bundle】-This Switch a...,[],23.99,{'hi_res': ['https://m.media-amazon.com/images...,"{'title': ['Switch Lite Case ', 'ProCase Ninte...",innoAura,"[Video Games, Nintendo Switch, Accessories, Ca...","{""Package Dimensions"": ""9.02 x 5.91 x 2.52 inc...",B0B2PR256F,NaN,NaN,NaN
4,Cell Phones & Accessories,"FANPL Design for Nintendo Switch OLED Case, Cu...",3.7,4,"[【Exclusive pattern】: FANPL pattern design, re...","[【Features】, ✔️FANPL pattern design shows more...",None,{'hi_res': ['https://m.media-amazon.com/images...,"{'title': [], 'url': [], 'user_id': []}",FANPL,"[Video Games, Nintendo Switch, Accessories, Ca...","{""Package Dimensions"": ""8.19 x 4.33 x 1.3 inch...",B09WTQKH6P,NaN,NaN,NaN


## 2. Missing Values

In [8]:
reviews_df.isnull().sum()

rating               0
title                0
text                 0
images               0
asin                 0
parent_asin          0
user_id              0
timestamp            0
helpful_vote         0
verified_purchase    0
dtype: int64

In [9]:
metadata_df.isnull().sum()

main_category        8102
title                   0
average_rating          0
rating_number           0
features                0
description             0
price                   0
images                  0
videos                  0
store                3186
categories              0
details                 0
parent_asin             0
bought_together    100000
subtitle            99740
author              99805
dtype: int64

### 3. Interaction Distribution

In [14]:
user_counts = reviews_df.groupby("user_id").size()
item_counts = reviews_df.groupby("parent_asin").size()

print(user_counts.describe())
print((user_counts == 1).mean())
print("########")
print(item_counts.describe())
print((item_counts == 1).mean())

count    2.193964e+06
mean     1.547308e+00
std      1.980513e+00
min      1.000000e+00
25%      1.000000e+00
50%      1.000000e+00
75%      1.000000e+00
max      5.000000e+02
dtype: float64
0.7521417853711364
########
count    99986.000000
mean        33.952123
std        205.014600
min          1.000000
25%          1.000000
50%          4.000000
75%         14.000000
max      18105.000000
dtype: float64
0.2770387854299602


**Users are extremely sparse:**
- 2,193,964 unique users, mean of 1.55 reviews/user
- 75.2% of users have exactly 1 review (median = 1, max = 500)
- This is typical of e-commerce data (most people review a product once, if at all)
  but means most users have almost no history to learn a preference from

**Items are richer but still long-tailed:**
- 99,986 unique items, mean of 33.95 reviews/item, but median is only 4
- 27.7% of items have only 1 review (max = 18,105 — a handful of extremely popular items
  dominate the distribution)
- The large gap between mean and median confirms a long-tail / power-law shape typical of Amazon data

**Decision: standard cutoffs (e.g. 5+ per user) don't apply here**
- A 5+ review cutoff would eliminate ~75%+ of users, which is too aggressive for this dataset
- Chose to filter with: `min_user_interactions = 2`, `min_item_interactions = 5`
- Filtering will be applied iteratively (removing low-count users can push some items
  below threshold, and vice versa) until the dataset stabilizes
- After filtering: [FILL IN — retained X reviews, Y users, Z items]

## 4. Rating Distribution

In [13]:
reviews_df["rating"].value_counts(normalize=True).sort_index()

rating
1.0    0.127700
2.0    0.054201
3.0    0.073563
4.0    0.133787
5.0    0.610748
Name: proportion, dtype: float64